### PR TITLE
cli: add trafficpolicy command to validate pods against traffic targets

### LINKED ISSUE #1945: Add a cli command to validate if 2 pods are allowed to communicate based on SMI TrafficTarget policies
**Please describe the Improvement and/or Feature Request**
This issue proposes to add a new cli command that will verify if 2 pods are allowed to communicate (direction matters) by going through the applied TrafficTarget SMI policies.

The command could look as follows:
```
$ osm trafficpolicy check --source-pod <src> --destination-pod <dst>
pod `src` is allowed to communicate to pod <dst> with SMI TrafficTarget policy <foo>
```

**Scope (please mark with X where applicable)**
- New Functionality          [ ]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [X]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [X]
- Tests                      [ ]
- CI System                  [ ]
- Project Release            [ ]

**Possible use cases**
If a TrafficTarget policy is misconfigured, this command would quickly detect this by matching the ServiceAccount in the pod spec to the one specified in the TrafficTarget policy.

### PR DESCRIPTION
**Pull request by shashankram**
**Description**:
This change adds a new cli command `osm trafficpolicy check-pods` that
validates whether the given source and destination pods are allowed
to communicate. It detects permissive mode vs SMI, and correctly
prints information regarding if the source pod is allowed to send
traffic to the destination pod.

Additionally, unit tests verify the different scenarios:
- permissive mode
- source and destination allowed by SMI
- source and destination not allowed by SMI

This command doesn't look into HTTP route policies to examine
individual HTTP routes. Extending this to routes requires more
thought.

Resolves #1945

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [X]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
`No`

### CODE REVIEW

File: cmd/cli/trafficpolicy_check.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+		return false, errors.Errorf("Missing key %q in %s/%s ConfigMap", permissiveTrafficPolicyModeKey, configMap.Namespace, configMap.Name)
+	}
+	configMapBoolValue, err := strconv.ParseBool(strval)
+	if err != nil {
+		return false, errors.Errorf("Invalid value %q for key %q in %s/%s ConfigMap, must be boolean", strval, permissiveTrafficPolicyModeKey, configMap.Namespace, configMap.Name)
+	}
+	return configMapBoolValue, nil
+}
+
+func unmarshalNamespacedPod(namespacedPod string) (string, string, error) {
```
Comment (draychev) [inline cmd/cli/trafficpolicy_check.go:193]: It is a little tough to understand what is being returned here with the 1 strings.

To make this easier to comprehend I would create explicit variable names that suggest what this is. For instance:
```go
namespace := chunk[0]
podName := chunk[1]
return namespace, podName
```

And of course we could use named return values - tells you what you get from this function at first glance.

File: cmd/cli/trafficpolicy_check.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	if !isMeshedPod(*pod) {
+		return nil, errors.Errorf("Pod %s in namespace %s is not a part of a mesh", podName, namespace)
+	}
+	return pod, nil
+}
+
+func (cmd *trafficPolicyCheckCmd) isPermissiveModeEnabled() (bool, error) {
+	osmNamespace := settings.Namespace()
+	configMap, err := cmd.clientSet.CoreV1().ConfigMaps(osmNamespace).Get(context.TODO(), osmConfigMapName, metav1.GetOptions{})
```
Comment (draychev) [inline cmd/cli/trafficpolicy_check.go:178]: We have an opportunity to reuse `pkg/configurator`
Do you think you could carve out this: https://github.com/openservicemesh/osm/blob/3af31b59f7dfaacb76e480f35cc9ae6cab2b44a1/pkg/configurator/client.go#L138-L156
into a separate function and use it here?

You already have `configMap` here - could pass it to a new function which contains the code from the link above and then use `.PermissiveTrafficPolicyMode`: https://github.com/openservicemesh/osm/blob/3af31b59f7dfaacb76e480f35cc9ae6cab2b44a1/pkg/configurator/methods.go#L41-L43

File: cmd/cli/trafficpolicy_check.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	trafficTargets, err := cmd.smiAccessClient.AccessV1alpha2().TrafficTargets(dstPod.Namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Errorf("Error listing SMI TrafficTarget policies: %s", err)
+	}
+
+	var foundTrafficTarget bool
+	for _, trafficTarget := range trafficTargets.Items {
+		spec := trafficTarget.Spec
+		// Map traffic targets to the given pods
+		if spec.Destination.Name == dstPod.Spec.ServiceAccountName && spec.Destination.Namespace == dstPod.Namespace {
```
Comment (draychev) [inline cmd/cli/trafficpolicy_check.go:135]: It seems to be that both for the destination and source below - we need to also constrain to `spec.Destination.Kind == "ServiceAccount"`

```yaml
kind: TrafficTarget
metadata:
  name: path-specific
  namespace: default
spec:
  destination:
    kind: ServiceAccount
    name: service-a
    namespace: default
    port: 8080
  rules:
  - kind: HTTPRouteGroup
    name: the-routes
    matches:
    - metrics
  sources:
  - kind: ServiceAccount
    name: prometheus
    namespace: default
```

**Review by draychev — APPROVED** [review]
This looks good. Will be a very useful tool.
I see a few opportunities to improve and re-use - but these could happen in another iteration of this PR.

File: cmd/cli/trafficpolicy.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+)
+
+const trafficPolicyDescription = `
+This command consists of subcommands related to traffic policies
+associated with osm.
+`
+
+func newTrafficPolicyCmd(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "trafficpolicy",
```
[the remark this answers is not in the document]
Comment (shashankram) [reply cmd/cli/trafficpolicy.go:16]: Sure, can shorten this to `policy`

File: cmd/cli/trafficpolicy_check.go (hunk 1 above)
Comment (shashankram) [reply cmd/cli/trafficpolicy_check.go:193]: named return is perfect for this tiny function. Will update.

File: cmd/cli/trafficpolicy_check.go (hunk 3 above)
Comment (shashankram) [reply cmd/cli/trafficpolicy_check.go:135]: Probably needs to happen in the controller as well.

File: cmd/cli/trafficpolicy_check.go (hunk 2 above)
Comment (shashankram) [reply cmd/cli/trafficpolicy_check.go:187]: This needs a bit more refactoring, created an issue to reuse code: https://github.com/openservicemesh/osm/issues/1962

File: cmd/cli/trafficpolicy.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+const trafficPolicyDescription = `
+This command consists of subcommands related to traffic policies
+associated with osm.
+`
+
+func newTrafficPolicyCmd(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "policy",
+		Short: "manage and check traffic policies",
```
Comment (snehachhabria) [inline cmd/cli/trafficpolicy.go:17]:
```suggestion
		Short: "manage and check SMI traffic policies",
```

File: cmd/cli/trafficpolicy_check.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+func (cmd *trafficPolicyCheckCmd) checkTrafficPolicy(srcPod, dstPod *corev1.Pod) error {
+	osmNamespace := settings.Namespace()
+
+	// Check if permissive mode is enabled, in which case every meshed pod is allowed to communicate with each other
+	if permissiveMode, err := cmd.isPermissiveModeEnabled(); err != nil {
+		return errors.Errorf("Error checking if permissive mode is enabled")
+	} else if permissiveMode {
+		fmt.Fprintf(cmd.out, "[+] Permissive mode enabled for mesh operated by osm-controller running in %s namespace\n\n "+
```
Comment (snehachhabria) [inline cmd/cli/trafficpolicy_check.go:119]:
```suggestion
		fmt.Fprintf(cmd.out, "[+] Permissive mode enabled for mesh operated by osm-controller running in '%s' namespace\n\n "+
```

**Review by snehachhabria — DISMISSED** [review]

File: cmd/cli/trafficpolicy.go (hunk 5 above)
Comment (shashankram) [reply cmd/cli/trafficpolicy.go:17]: Not just related to SMI, handles permissive mode as well.
